### PR TITLE
 mds: optimize the way how max export size is enforced

### PR DESCRIPTION
**Pull request by ukernel**

### CODE REVIEW

**Comment by tchaikov** [comment]
retest this please.

File: src/mds/Migrator.cc [hunk 1]
```diff
@@ -947,23 +1106,91 @@ void Migrator::dispatch_export_dir(MDRequestRef& mdr, int count)
...
-  // start the freeze, but hold it up with an auth_pin.
-  dir->freeze_tree();
-  assert(dir->is_freezing_tree());
-  dir->add_waiter(CDir::WAIT_FROZEN, new C_MDC_ExportFreeze(this, dir, it->second.tid));
+
+  vector<pair<CDir*, size_t> > results;
+  maybe_split_export(dir, results);
+
+  auto now = ceph_clock_now();
+  if (results.size() == 1 && results.front().first == dir) {
```
Comment (batrick) [inline src/mds/Migrator.cc]: What circumstance would `results.front().first == dir` not be true if `results.size() == 1`?

File: src/mds/Migrator.cc [hunk 2]
```diff
@@ -947,23 +1106,91 @@ void Migrator::dispatch_export_dir(MDRequestRef& mdr, int count)
...
+
+  auto parent = std::make_shared<export_base_t>(dir->dirfrag(), dest,
+						results.size(),
+						export_queue_gen);
+
+  dout(7) << "subtree is too large, splitting it into: " <<  dendl;
+  for (auto& p : results) {
+    CDir *sub = p.first;
+    assert(sub != dir);
+    dout(7) << " segment " << *sub << dendl;
```
Comment (batrick) [inline src/mds/Migrator.cc]: how about ` subtree split segment` to make it easier to find in logs.

File: src/mds/Migrator.cc [hunk 3]
```diff
@@ -947,23 +1106,91 @@ void Migrator::dispatch_export_dir(MDRequestRef& mdr, int count)
...
+							  mds->get_nodeid(), stat.tid);
+    mds->send_message_mds(discover, dest);
+
+    // start the freeze, but hold it up with an auth_pin.
+    sub->freeze_tree();
+    assert(sub->is_freezing_tree());
+    sub->add_waiter(CDir::WAIT_FROZEN, new C_MDC_ExportFreeze(this, sub, stat.tid));
+  }
+
+  // cancel the origin one
```
Comment (batrick) [inline src/mds/Migrator.cc]: *original

File: src/mds/Mutation.h [hunk 4]
```diff
@@ -222,66 +222,58 @@ struct MDRequestImpl : public MutationImpl {
   // break rarely-used fields into a separately allocated structure 
   // to save memory for most ops
   struct More {
-    int slave_error;
+    int slave_error = 0;
```
Comment (batrick) [inline src/mds/Mutation.h]: Please make these types of changes a separate commit

Comment (batrick) [reply src/mds/Mutation.h]: (these changes obscure that you added `export_ref`)

File: src/mds/Migrator.cc [hunk 5]
```diff
@@ -867,27 +871,181 @@ void Migrator::export_dir(CDir *dir, mds_rank_t dest)
...
+    if (data.complete) {
+      auto cur_size = data.subdirs_size + dirfrag_size;
+      // we can do nothing with large dirfrag
+      if (cur_size >= max_size && found_size * 2 > max_size)
+	break;
+
+      found_size += dirfrag_size;
+
+      if (stack.size() > 1) {
+	auto& parent = stack[stack.size() - 2];
```
[the remark this answers is not in the document]
Comment (ukernel) [reply src/mds/Migrator.cc]: already checked size, [] can avoid validity of at

File: src/mds/Migrator.cc (hunk 1 above)
Comment (ukernel) [reply src/mds/Migrator.cc]: original subtree A
[pre]
   A
/    \
B    C
[/pre]

maybe_split_export only wants to export subtree B (subtree B is large enough)

File: src/mds/Migrator.cc [hunk 6]
```diff
@@ -2155,10 +2332,14 @@ void Migrator::handle_export_discover(MExportDirDiscover *m)
     vector<CDentry*> trace;
     MDRequestRef null_ref;
     int r = cache->path_traverse(null_ref, m, NULL, fpath, &trace, NULL, MDS_TRAVERSE_DISCOVER);
-    if (r > 0) return;
+    if (r > 0)
+      return;
```
[the remark this answers is not in the document]
Comment (ukernel) [reply src/mds/Migrator.cc]: "r > 0" means the message is added to C_MDS_RetryMessage

**Comment by ukernel** [comment]
@batrick updated

**Comment by batrick** [comment]
@ukernel needs tracker tickets still for backport?

**Comment by batrick** [comment]
@ukernel would it make sense to test this (outside of teuthology) by creating a directory with a ~million dentries and trying to pin the directory?

**Comment by ukernel** [comment]
Yes.  Also need to make a client cache most of the files. The slowest part of subtree migration is send cap export messages.

**Comment by ukernel** [comment]
http://tracker.ceph.com/issues/26867 should be fixed by the last commit. I also update "mds: optimize the way how max export size is enforced", make it not change export locking

**Comment by batrick** [comment]
@ukernel FYI I'm planning to review this tomorrow or Friday to make this weekend's test run.

**Comment by batrick** [comment]
This needs rebased to handle new message handling introduced by #22555.
